### PR TITLE
8322815: Eliminate -Wparentheses warnings in shenandoah code

### PR DESCRIPTION
**Pull request by kimbarrett**
8322815: Eliminate -Wparentheses warnings in shenandoah code

Please review this change to eliminate some -Wparentheses warnings. This
involved simply adding a few parentheses to make some implicit operator
precedence explicit.

Testing: local (linux-x64) build with shenandoah included and ran
hotspot:tier1 with -XX:+UseShenandoahGC.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322815](https://bugs.openjdk.org/browse/JDK-8322815): Eliminate -Wparentheses warnings in shenandoah code (**Enhancement** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [b0d9d50d](https://git.openjdk.org/jdk/pull/17214/files/b0d9d50dc1ff6aaa836726db944e7c8ca27bf8fa)
 * [Zhengyu Gu](https://openjdk.org/census#zgu) (@zhengyu123 - **Reviewer**) ⚠️ Review applies to [b0d9d50d](https://git.openjdk.org/jdk/pull/17214/files/b0d9d50dc1ff6aaa836726db944e7c8ca27bf8fa)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17214/head:pull/17214` \
`$ git checkout pull/17214`

Update a local copy of the PR: \
`$ git checkout pull/17214` \
`$ git pull https://git.openjdk.org/jdk.git pull/17214/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17214`

View PR using the GUI difftool: \
`$ git pr show -t 17214`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17214.diff">https://git.openjdk.org/jdk/pull/17214.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17214#issuecomment-1873649124)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp [hunk 1]
```diff
@@ -707,7 +707,7 @@ void ShenandoahEvacUpdateCleanupOopStorageRootsClosure::do_oop(oop* p) {
       }
       ShenandoahHeap::atomic_update_oop(resolved, p, obj);
       assert(_heap->cancelled_gc() ||
-             _mark_context->is_marked(resolved) && !_heap->in_collection_set(resolved),
+             (_mark_context->is_marked(resolved) && !_heap->in_collection_set(resolved)),
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:710]: The change for this PR seems fine, but the whole assert looks weird, and lacks richer diagnostics Shenandoah asserts do. I believe the forwardee is explicitly marked, given it is a new copy after TAMS. So the whole thing might be just:

```
  shenandoah_assert_not_in_cset_except(p, resolved, _heap->cancelled_gc());
```

Right, @zhengyu123? We should do that in separate PR.

**Review by shipilev — APPROVED** [review]

**Review by zhengyu123 — APPROVED** [review]
LGTM

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp]: Thanks.  Removed as part of merge conflict resolution.

**Comment by kimbarrett** [comment]
Thanks for reviews @shipilev and @zhengyu123 .
